services/wireguard/endpoint.go

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright (C) 2018 The "MysteriumNetwork/node" Authors.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package wireguard
+
+import (
+	"net"
+	"os/exec"
+
+	"github.com/mdlayher/wireguardctrl"
+	"github.com/mdlayher/wireguardctrl/wgtypes"
+	"github.com/mysteriumnetwork/node/core/ip"
+	"github.com/mysteriumnetwork/node/services/wireguard/resources"
+)
+
+type connectionEndpoint struct {
+	iface      string
+	ip         net.IPNet
+	endpoint   net.UDPAddr
+	wgClient   *wireguardctrl.Client
+	resources  resources.Handler
+	ipResolver ip.Resolver
+}
+
+// NewConnectionEndpoint creates Wireguard client with predefined interface name and endpoint.
+func NewConnectionEndpoint(ipResolver ip.Resolver) (*connectionEndpoint, error) {
+	wgClient, err := wireguardctrl.New()
+	if err != nil {
+		return nil, err
+	}
+
+	return &connectionEndpoint{
+		wgClient:   wgClient,
+		ipResolver: ipResolver,
+		resources:  resources.Handler{},
+	}, nil
+}
+
+// Start starts and configure wireguard network interface for providing service.
+func (ce *connectionEndpoint) Start() error {
+	publicIP, err := ce.ipResolver.GetPublicIP()
+	if err != nil {
+		return err
+	}
+
+	ce.ip = ce.resources.AllocateIP()
+	ce.iface = ce.resources.AllocateInterface()
+	ce.endpoint = net.UDPAddr{IP: net.ParseIP(publicIP), Port: ce.resources.AllocatePort()}
+	if err := ce.up(); err != nil {
+		return err
+	}
+
+	key, err := wgtypes.GeneratePrivateKey()
+	if err != nil {
+		return err
+	}
+
+	if err := ce.wgClient.ConfigureDevice(ce.iface, wgtypes.Config{
+		PrivateKey:   &key,
+		ListenPort:   &ce.endpoint.Port,
+		ReplacePeers: true,
+	}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// NewConsumer adds service consumer public key to the list of allowed peers.
+func (ce *connectionEndpoint) NewConsumer() (configProvider, error) {
+	peerKey, err := wgtypes.GeneratePrivateKey()
+	if err != nil {
+		return consumer{}, err
+	}
+
+	ip := ce.resources.AllocateIP()
+	config := wgtypes.Config{
+		Peers: []wgtypes.PeerConfig{{
+			Endpoint:   &ce.endpoint,
+			PublicKey:  peerKey.PublicKey(),
+			AllowedIPs: []net.IPNet{ip},
+		}},
+	}
+
+	if err := ce.wgClient.ConfigureDevice(ce.iface, config); err != nil {
+		return consumer{}, err
+	}
+
+	return consumer{
+		ip:         ip,
```

Comment:
After discussing with @tadovas we may call it either tunnelAddress or interfaceAddress

Reply:
tunnelAddress sounds strange, tunnel is point to point and have 2 addresses commonly. interfaceAddress sounds like interface hardware address. Isn’t consume.ip better choice?

Reply:
ip is too generic. Lets use WG terms. WG does not have such notion. It is always good to relate terms with ones used in documentation (manual). Otherwise it takes longer to grasp its usage.

Reply:
> ip is too generic. 
btw, it not just ip, it's always `consumer.ip`, what can it mean except IP address assigned to consumer?